backends/p4tools/testgen/targets/bmv2/p4_asserts_parser.cpp

```diff
@@ -0,0 +1,623 @@
+            while (flag) {
+                i++;
+                if (tokens[i].kind() == Token::Kind::LeftParen) {
+                    first = false;
+                    continue;
+                } else if (first) {
+                    break;
+                }
+                if (tokens[i].kind() != Token::Kind::RightParen) {
+                    str += static_cast<std::string>(tokens[i].lexeme());
+                } else {
+                    flag = false;
+                }
+            }
+            i = idx;
+            if (str.size() == 0) {
+                str = "tmp";
+            }
+            const IR::Expression* expr1 = new IR::PathExpression(new IR::Path(IR::ID(str)));
+            exprVec.push_back(new IR::LNot(expr1));
+        } else if (tokens[i].kind() == Token::Kind::Implication) {
+            auto tmp = exprVec[exprVec.size() - 1];
+            exprVec.pop_back();
+            const IR::Expression* expr1 = new IR::PathExpression(new IR::Path(IR::ID("tmp")));
+            exprVec.push_back(new IR::LNot(expr1));
+            exprVec.push_back(tmp);
+            const IR::Expression* expr2 = new IR::PathExpression(new IR::Path(IR::ID("(tmp")));
+            exprVec.push_back(new IR::LOr(expr1, expr2));
+        } else if (tokens[i].kind() == Token::Kind::Disjunction) {
+            const IR::Expression* expr1 = new IR::PathExpression(new IR::Path(IR::ID("tmp")));
+            const IR::Expression* expr2 = new IR::PathExpression(new IR::Path(IR::ID("tmp")));
+            exprVec.push_back(new IR::LOr(expr1, expr2));
+        } else if (tokens[i].kind() == Token::Kind::Conjunction) {
+            const IR::Expression* expr1 = new IR::PathExpression(new IR::Path(IR::ID("tmp")));
+            const IR::Expression* expr2 = new IR::PathExpression(new IR::Path(IR::ID("tmp")));
+            exprVec.push_back(new IR::LAnd(expr1, expr2));
+        }
+    }
+    std::vector<const IR::Expression*> tmp;
+    for (long unsigned int i = 0; i < exprVec.size(); i++) {
+        if (auto lNot = exprVec[i]->to<IR::LNot>()) {
+            i++;
+            auto lNotStr = lNot->expr->toString();
+
+            auto binary = exprVec[i]->checkedTo<IR::Operation_Binary>()->clone();
+            if (auto subBin = binary->left->to<IR::Operation_Binary>()) {
+                cstring str = subBin->getStringOp();
+                str.replace(" ", "");
+
+                if (lNotStr.find(str)) {
+                    binary->left = new IR::LNot(binary->left);
+                    tmp.push_back(binary);
+                    continue;
+                }
+            }
+
+            tmp.push_back(new IR::LNot(exprVec[i]));
+            continue;
+        } else {
+            tmp.push_back(exprVec[i]);
+        }
+    }
+    if (tmp.size() == 1) {
+        return expr;
+    }
+
+    expr = makeExpr(tmp);
+
+    return expr;
+}
+
+std::vector<const IR::Expression*> AssertsParser::genIRStructs(cstring str) {
+    Lexer lex(str);
```

Comment:
Similar comments apply to all these other functions. They are huge and very hard to read.